File: frontend/src/services/matchingService.js
import api from '../api/axios';

/**
 * Service managing blood matching engine endpoints.
 */
export const matchingService = {
  /**
   * Retrieves a ranked list of eligible, compatible donors for a verified request.
   */
  getRankedDonors: async (requestId) => {
    const response = await api.get(`/matching/request/${requestId}`);
    return response.data;
  },

  /**
   * Generates and persists matching records for a blood request.
   */
  generateMatches: async (requestId) => {
    const response = await api.post(`/matching/request/${requestId}/generate`);
    return response.data;
  },

  /**
   * Retrieves generated match records for a request.
   */
  getGeneratedMatches: async (requestId) => {
    const response = await api.get(`/matching/request/${requestId}/results`);
    return response.data;
  },
};

export default matchingService;

File: backend/src/main/java/com/bloodbridge/service/MatchingService.java
package com.bloodbridge.service;

import com.bloodbridge.dto.response.ApiResponse;
import com.bloodbridge.dto.response.MatchDashboardResponse;
import com.bloodbridge.dto.response.MatchResponse;

import java.util.List;

/**
 * Service interface for Intelligent Blood Matching & Allocation Engine workflows.
 */
public interface MatchingService {

    /**
     * Triggers the matching engine for a specific blood request, persists MatchResult entities, and returns top ranked matches.
     *
     * @param bloodRequestId target blood request ID
     * @return ApiResponse containing list of MatchResponse items
     */
    ApiResponse<List<MatchResponse>> triggerMatching(Long bloodRequestId);

    /**
     * Retrieves existing match results for a blood request ordered by rank.
     *
     * @param bloodRequestId blood request ID
     * @return ApiResponse containing list of MatchResponse items
     */
    ApiResponse<List<MatchResponse>> getMatchesByRequestId(Long bloodRequestId);

    /**
     * Retrieves all match results across requests.
     *
     * @return ApiResponse containing list of MatchResponse items
     */
    ApiResponse<List<MatchResponse>> getAllMatches();

    /**
     * Recalculates and replaces matches for a blood request (e.g. when availability or request details change).
     *
     * @param bloodRequestId blood request ID
     * @return ApiResponse containing list of updated MatchResponse items
     */
    ApiResponse<List<MatchResponse>> recalculateMatches(Long bloodRequestId);

    /**
     * Retrieves hospital match dashboard overview for a blood request.
     *
     * @param bloodRequestId blood request ID
     * @return ApiResponse containing MatchDashboardResponse
     */
    ApiResponse<MatchDashboardResponse> getMatchDashboard(Long bloodRequestId);
}
